### DISCUSSION TITLE
Is there a way to create a nested path instead of a subdocument?

### DISCUSSION THREAD

**Discussion by eric-swann-q2** [typegoose/typegoose#612] [Q&A]
Is it possible to create a nested path as opposed to a subdocument and have tracking etc.. function? 

In my case, I'm converting complex nested schemas to Typegoose, but am being forced to change what were previously basic nested paths into many separate nested subdocuments. I would be OK with this, but there are behavioral differences of subdocuments versus nested path in mongoose (I'm sure you know this already). It looks like Typegoose requires all nested entities to be subdocuments. 

The main behavioral difference is that that a nested path is inherently lean, while a nested document comes with it's own wrapper and tracking, which is unnecessary in the vast majority of cases and messes up a lot of existing document parsing logic that we have in place on our project.

On the flip side, if I create a basic nested path inside of the parent, it doesn't track property changes with Typegoose.

Is there a way to force a nested entity to be a simple nested path and still function properly with Typegoose?

If not, is this something you would consider adding as a custom decorator?

Thanks,

## Versions
I am running 8.3.0 on NodeJS 14 with Typescript 4.4.3. Mongoose is 5.13.8.

**Comment by hasezoey** [comment]
yes there is a way in typegoose, called the [`Passthrough` class](https://typegoose.github.io/typegoose/docs/api/decorators/prop/#passthrough-class), though it is not recommended to be used, when otherwise possible

also starting with mongoose 6.0.0 (and so typegoose 9.0.0), both ways should be equal (see https://github.com/Automattic/mongoose/issues/7181)

**Reply by eric-swann-q2** [reply]
Maybe I'm misunderstanding, but this seems like a separate issue.  
Mongoose 6 fully supports both subdocuments and nested paths: (https://mongoosejs.com/docs/subdocs.html#subdocuments-versus-nested-paths). I hope I'm wrong, but Typegoose only appears to support subdocuments, which has different behavior.

I cannot simply nest a substructure in the schema (as you can easily do in Mongoose) as the TS decorators do not allow it. So the only alternative is to nest a bunch of subdocument types.  So the following works fine as a nested path in Mongoose, including Mongoose 6, but you can see this doesn't work in Typegoose. It seems that Typegoose doesn't support simple nested paths. 
![image](https://user-images.githubusercontent.com/13240703/133866972-b7311c44-21d2-43d3-8d63-9cabb28dd5bf.png)
The only alternative is to make a full subdocument with it's own schema, and this has significant behavioral differences as mentioned.

**Reply by hasezoey** [reply]
> I hope I'm wrong, but Typegoose only appears to support subdocuments, which has different behavior.

like i said earlier, you can use the [`Passthrough` class](https://typegoose.github.io/typegoose/docs/api/decorators/prop/#passthrough-class) that typegoose provides, but it is not recommended to be used, because the paths will result in mixed (see [this documentation from mongoose](https://mongoosejs.com/docs/subdocs.html#altsyntaxsingle))

> It seems that Typegoose doesn't support simple nested paths.

it is true that typegoose cannot support paths like:
```ts
class Test {
  @prop()
  public nested: {
    prop1: string;
  }
}
```
the reason being, that this object will not be in the reflection, and so cannot be parsed at runtime, that is why in this case the `Passthrough` class has to be used

**Reply by eric-swann-q2** [reply]
OK, thanks for the info.

**Reply by hasezoey** [reply]
update: if you want to use sub-paths (not sub-documents) also not using `type` inbetween, the `Passthrough` class gets this feature in 9.0.0 (commit 73798104c8865e80ababba1663946fda6fefdb11)

**Reply by JanKalkan** [reply]
@hasezoey Catching up on this topic. 
[1] If I understand it correctly, the only way to be able to use nested paths in Typegoose is by using the Passthrough class which in turn always results in a mixed type in Mongoose, correct? 
[2] The actual consequence would be that Mongoose can't track changes on props defined in nested paths anymore, right? The Typegoose typing should still work even when the schema got populated?
[3] Any plans or chances to getting a proper solution in the near future? We are migrating our existing code base from JS to TS and have quite a lot of nested paths. That could potentially be the blocker for us.

**Reply by hasezoey** [reply]
@JanKalkan 

> If I understand it correctly, the only way to be able to use nested paths in Typegoose is by using the Passthrough class which in turn always results in a mixed type in Mongoose, correct?

no you are not understanding correctly, yes typegoose does `new Schema({ yourPath: { type: new Schema({ prop: { type: String } }) });`, instead of `new Schema({ yourPath: { type: { prop: { type: string } } })`, which since mongoose 6.0.0 [brings the same result](https://github.com/Automattic/mongoose/issues/7181).

The `Passthrough` class is meant to literally pass-through its arguments as if it would be a normal `new Schema({ HERE })` definition on the defined property (with the differentiation of `direct` and `non-direct`), see [the Passthrough-Class documentation](https://typegoose.github.io/typegoose/docs/api/decorators/prop#passthrough-class).

TL;DR: the `Passthrough`-class allows you to overwrite what typegoose would normally do, and it is encouraged to not be used (and if the need arises, to open a issue here)

> the only way to be able to use nested paths in Typegoose

as for how to do nesting, please see the first example in [the quick-start-guide](https://typegoose.github.io/typegoose/docs/guides/quick-start-guide) with property `mainJob` for direct nesting and `jobs` for array of nested, or the [Nesting Classes](https://typegoose.github.io/typegoose/docs/guides/nesting-classes) guide (which also contains a example of how to use the `Passthrough`-class).

**Reply by JanKalkan** [reply]
Still not sure if we are talking about the same thing. I am referring to a nested path in terms how Mongoose defines it [here](https://mongoosejs.com/docs/6.x/docs/subdocs.html#subdocuments-versus-nested-paths). 

```
// Nested path example
const nestedSchema = new mongoose.Schema({
  child: { name: String, age: Number }
});
```
In this case child is just a prop and not a subdocument and thus doesn't get an _id autogenerated etc. What would be the best approach to port this example to Typegoose and are there any limitations to keep in mind? Sorry for any confusion!

**Reply by hasezoey** [reply]
> I am referring to a nested path in terms how Mongoose defines it [here](https://mongoosejs.com/docs/6.x/docs/subdocs.html#subdocuments-versus-nested-paths).

i think the documentation is wrong, see https://github.com/Automattic/mongoose/issues/7181, https://github.com/Automattic/mongoose/commit/a51fbcc3442204895ae97e68b6d547be8d4b2f5c, https://github.com/Automattic/mongoose/commit/ec1b60166f091fcc54c4a0a7ce08ef47ec58030b; at least to my understanding (created https://github.com/Automattic/mongoose/issues/13272)

> In this case child is just a prop and not a subdocument and thus doesn't get an _id autogenerated etc

you can disable `_id` generation for subdocuments with either the [prop option `_id`](https://typegoose.github.io/typegoose/docs/api/decorators/prop#_id) or [the schema option `_id`](https://mongoosejs.com/docs/guide.html#_id)

> What would be the best approach to port this example to Typegoose and are there any limitations to keep in mind?

```ts
@modelOptions({ schemaOptions: { _id: false } }) // via the schema option
class Child {
  @prop()
  public name: string;

  @prop()
  public age: number;
}

class Nested {
  @prop({ _id: false }) // via the prop option
  public child: Child;
}
```

and no limitations that i know of

**Reply by AChangXD** [reply]
@JanKalkan I did something like 
```ts
class Child{
readonly _id: string;
@prop({required: true, unique: true})
prop1: string;

@prop()
prop2: boolean;
}

class Nested{
readonly _id: string;

@prop({type: Child})
child: Child;
}
```

My only issue so far is that the unique index I set in the child got carried over to the parent :(

**Reply by hasezoey** [reply]
https://github.com/Automattic/mongoose/issues/13272 has concluded and there still is a difference between `prop: { type: { anotherProp: String } }` and `prop: { type: new Schema({ anotherProp: String }) }`, and if i understood the conclusion correctly, the `new Schema` way provides some more functionality (which is what typegoose is currently doing) - though yes, some thing like automatic `_id` is inherent in this way and needs to be manually disabled
